source/DoorNode.cpp

```diff
@@ -25,26 +25,49 @@ int frameCap = 0;
 
 void DoorNode::draw(const std::shared_ptr<cugl::SpriteBatch>& batch, const Mat4& transform,
 					Color4 tint) {
-	Vec2 doorPos = Vec2(DOOR_RADIUS * sin(doorModel->getAngle() * globals::PI_180),
-						-(DOOR_RADIUS)*cos(doorModel->getAngle() * globals::PI_180));
-	if (doorModel->getAngle() < 0) {
-		doorPos = Vec2(OFF_SCREEN_POS, OFF_SCREEN_POS);
-	}
-	setPosition(doorPos);
-	setAngle(doorModel->getAngle() * globals::PI_180);
-
-	frameCap = doorModel->getPlayersOn() < 2 ? doorModel->getPlayersOn() * ONE_PLAYER_FRAME
-											 : TWO_PLAYER_FRAME;
-	if (getFrame() < frameCap) {
-		setFrame((int)getFrame() + 1);
-	} else if (getFrame() > frameCap) {
-		setFrame((int)getFrame() - 1);
-	}
+	Vec2 doorPos;
+	if (doorModel->getAngle() >= 0) {
+		// Door is currently active
+		float onScreenAngle = doorModel->getAngle() - playerDonutModel->getAngle();
+		onScreenAngle = onScreenAngle >= 0 ? onScreenAngle : shipSize + onScreenAngle;
+		onScreenAngle = onScreenAngle > shipSize / 2 ? onScreenAngle - shipSize : onScreenAngle;
+		onScreenAngle *= globals::PI_180;
+		if (!isShown && onScreenAngle < globals::SEG_CUTOFF_ANGLE &&
+			onScreenAngle > -globals::SEG_CUTOFF_ANGLE) {
+			// Door is coming into visible range
+			float relativeAngle = onScreenAngle - getParent()->getParent()->getAngle();
+			doorPos = Vec2(DOOR_RADIUS * sin(relativeAngle), -DOOR_RADIUS * cos(relativeAngle));
+			setPosition(doorPos);
+			isShown = true;
+			setAngle(relativeAngle);
+			CULog("Door coming into view at %f", onScreenAngle / globals::PI_180);
+		} else if (isShown && (onScreenAngle >= globals::SEG_CUTOFF_ANGLE ||
+							   onScreenAngle <= -globals::SEG_CUTOFF_ANGLE)) {
+			// Door is leaving visible range
+			doorPos = Vec2(OFF_SCREEN_POS, OFF_SCREEN_POS);
+			setPosition(doorPos);
+			isShown = false;
+			CULog("Door going out at %f", onScreenAngle / globals::PI_180);
+		}
```

Comment:
This is not urgent for this sprint, but I'm seeing this block of code a lot among all the nodes; is it possible to factor it out somehow?